### PR TITLE
util/ExportToCsv: Only duplicate media on export when they are unique

### LINKED ISSUE #77: Only duplicate media on export when there are truly unique
Collect uses timestamps to name files. On a very large campaign, two devices can take images at the same time, upload them both to Aggregate, then download them to Briefcase. When those images are exported, what happens?

### PR DESCRIPTION
**Pull request by shivam-tripathi**
While exporting, if the destination folder already contains a file with the
same name, compute the MD5 hash of the files and compare them. If
same, skip copying; else append a suffix to differentiate the files.
For better performance, cache the already computed hash in a HashMap with
fileName-hashValue as key-value pair.

Fix #77

### CODE REVIEW

File: src/org/opendatakit/briefcase/util/ExportToCsv.java [hunk 1]
```diff
@@ -306,6 +308,58 @@ private String getSubmissionValue(EncryptionInformation ei, TreeElement model, E
...
+    StringBuffer sb = new StringBuffer();
+    for (int i = 0; i < mdbytes.length; i++) {
+      sb.append(Integer.toString((mdbytes[i] & 0xff) + 0x100, 16).substring(1));
+    }
+
+    String hashVal = sb.toString();
+    fileHashMap.put(fileLocation, hashVal);
+
+    return hashVal;
+  }
```
Comment (icemc) [inline src/org/opendatakit/briefcase/util/ExportToCsv.java]: @shivam-tripathi There is already a method in briefcase/util known as getMD5hash that returns the MD5 hash of a file. i think you should make use of that. Cheers

Comment (shivam-tripathi) [inline src/org/opendatakit/briefcase/util/ExportToCsv.java]: Yeah! That surely skipped me. Thanks @icemc! :smile:

**Comment by shivam-tripathi** [comment]
Presently focusing only on code review, if there is a problem with removal of trailing spaces - will make a separate commit/ignore them later.

**Comment by yanokwa** [comment]
@shivam-tripathi How did you test that this does what you intended?

**Comment by shivam-tripathi** [comment]
Export any form with redundant media files, say for instance Birds. If further testing is needed, one can make a new form with needed entries.

**Comment by yanokwa** [comment]
Hi @shivam-tripathi! Thanks again for sending in the PR. 

It's important to be thorough and precise in how you describe your tests. I know this takes time, but skipping it doesn't save time, it just shifts the burden to the maintainers. This is why some projects reject PRs that don't have testing descriptions or tests included because it's probable those PRs have bugs.

I tested your PR against the situation described in #77. 
> Collect uses timestamps to name files. On a very large campaign, two devices can take images at the same time, upload them both to Aggregate, then download them to Briefcase. When those images are exported, what happens?

To re-create this situation, I created an ODK Briefcase Storage folder with two submissions. The differences between the submissions are:
* UUID in submission.xml is different. One ends in `a0d1`, the other ends in `a0d9`. And because the UUID is different, the folder name is different.
* The image has the same name, but a different size because one has been compressed. Basically, they are different images.

screen shot 2017-04-13 at 12 39 59

When I ran your PR and exported the data with the GUI, I only got one image (the small one) in the media folder. The CSV also had two rows that pointed to the same file. I was expecting two image files  (`1490380664332.jpg` and `1490380664332-2.jpg`) and a CSV with rows that pointed to the appropriate files.

I did not test the CLI, but I bet the behavior is the same. I also did not test what happens if I added a third submission with a different file size.

Can you please fix this bug? And when you update the PR, please leave a comment with the details of your testing. An ideal PR should not require me to run the code. I mean, I will run the code, but I should be able to read your description of the testing and verify that you've thought of all the things.

Thanks again for the PR, and I'm looking forward to the fix!

**Comment by shivam-tripathi** [comment]
@yanokwa I am having a little trouble with the hardware, my apologies for not being thorough. I will be unable to write the fix (and test it) for the same reason for few more days (two to three at best). I hope that would be fine. 
Thanks for taking out time and highlighting important things. I will work on it as soon as possible. I was working on writing a complete description about the test, but I guess you beat me at it!
The issue you mentioned is probably because of the MD5 hash being generated same for both the files. It appears very strange, as I believe they should not be same. I will go again through the code for generating the MD5 hash and get back to you as soon as possible.
Again, my apologies for not being thorough.

**Comment by shivam-tripathi** [comment]
@yanokwa I am having a little trouble with the hardware, my apologies for not being thorough.
Thanks for taking out time and highlighting important things. I will work on it as soon as possible. I was working on writing a complete description about the test, but I guess you beat me at it!
I checked and found out the reason for the issue. I made a mistake.
Again, my apologies for not being thorough.

**Comment by shivam-tripathi** [comment]
@yanokwa I am having a little trouble with the hardware, my apologies for not being thorough.
Thanks for taking out time and highlighting important things. I will work on it as soon as possible. I was working on writing a complete description about the test, but I guess you beat me at it!
I checked and found out the reason for the issue. 
Again, my apologies for not being thorough.

**Pull request closed without merging**